velox/expression/Expr.cpp

```diff
@@ -420,14 +420,25 @@ bool Expr::checkGetSharedSubexprValues(
     assert(missingRows); // lint
     missingRows->deselect(*sharedSubexprRows_);
 
-    // Fix finalSelection at "rows" if missingRows is a strict subset to avoid
-    // losing values outside of missingRows.
-    bool updateFinalSelection = context.isFinalSelection() &&
-        (missingRows->countSelected() < rows.countSelected());
-    VarSetter finalSelectionOr(
-        context.mutableFinalSelection(), &rows, updateFinalSelection);
-    VarSetter isFinalSelectionOr(
-        context.mutableIsFinalSelection(), false, updateFinalSelection);
+    // Add the missingRows to sharedSubexprRows_ that will eventually be
+    // evaluated and added to sharedSubexprValues_.
+    sharedSubexprRows_->select(*missingRows);
+
+    // Fix finalSelection to avoid losing values outside missingRows.
+    LocalSelectivityVector newFinalSelectionHolder(
+        context, *sharedSubexprRows_);
+    auto newFinalSelection = newFinalSelectionHolder.get();
+    assert(newFinalSelection); // lint
+    if (context.finalSelection() != nullptr) {
```

Comment:
Shouldn't this check be !context.isFinalSelection() ?

Reply:
I noticed that there is an instance where isFinalSelection is not set to false but finalSelection is update, like in And conjunct so I am checking directly for that to avoid overriding finalSelection in case isFinalSelection does get set to false in any of the children of this expression.

Reply:
Not sure I understand. EvalCtx::isFinalSelection determines whether the value in EvalCtx::finalSelection is meaningful or not. Only when isFinalSelection == false, we should look at finalSelection. Hence, I think the check should use isFinalSelection, not finalSelection. If it happens to be that isFinalSelection == false, but finalSelection is null, then it is a bug.

Reply:
I was referring to this part of the code: https://github.com/facebookincubator/velox/blob/main/velox/expression/ConjunctExpr.cpp#L140-L143.

However, digging a bit into it now , I noticed that it is being overridden anyways whenever finalSelection is activated (isFinalSelection set to false), so we should be able to remove that part and tie both isFinalSelection+finalSelection in a single state transition and also implement your suggestion in https://github.com/facebookincubator/velox/pull/2399#discussion_r958936691).

I'll go ahead and update this to check !context.isFinalSelection() as well.